### DISCUSSION TITLE
How to properly access variables and aux variables at arbitrary nodes

### DISCUSSION THREAD

**Discussion by Eilloo** [idaholab/moose#23236] [Q&A General] [answered]
Hi all,

I'm wondering what the 'best' way to access both variables, and aux variables at some arbitrary node is - for example, at the nearest node on a particular boundary, or a neighbour to the current node. Obtaining the node ID's is not a problem.

I have had some success by setting the coupled variable type to be an integer in the header file, then using this to to get the dof number for that variable at a particular node, and finally pulling the appropriate value out from _nl.sys.currentSolution.

This was an approach I saw in another auxkernel (I'd have to dig around to remember which one!), rather than something I came up with myself.

The problem I am now having is that it doesn't seem to work for aux variables?
When I try and pass in aux variables using the same syntax, the integer associated with them is negative. Even stranger to me, the value doesn't change whether I include the aux variable in the input file or not (it is not a required variable). 
All of this makes me think the negative integer values I'm seeing associated with aux variables here are nonsense?

Now I'm wondering two things:

1) Are aux variables also stored in currentSolution, or are they dealt with differently? Is there a way I can access their values at a specific node?

2) Is there a conventional way of doing this in general? So far I have been using a method from another aux kernel, but perhaps this isn't the intended/best method for achieving this?

Hopefully the problem makes sense - if anything needs clarification, please let me know.
Thanks!

**Comment by GiudGiud** [comment]
Hello

auxiliary variables should be in the auxiliary system solution not the _nl

I can look for them later but this has been a fairly common need and there s at least another post about this on this forum.

Guillaume

**Reply by Eilloo** [reply]
Hi Guillaume, thanks for the quick response!

I have been looking on the forum but so far haven't found anything looking specifically at aux variables at nodes besides the current node. If you do happen to know of any threads I should take a look at, these are very welcome (and apologies if this turns out to be a duplicate!).

Thanks for pointing out the different systems - I'll look at the auxiliary system documentation to try and get the right syntax.

Negative integers (in my head, I'm expecting these to represent an index?) associated with the aux variables still seems strange, but I could be misunderstanding what's going on behind the scenes, and what these values represent? 
Even stranger is the fact they don't change, whether I do or don't pass in a variable from the input file, so I think I must be totally going wrong in this part.

Just in case I've done something daft in the code, I have `const int _myAuxVar` in the header file, then `params.addCoupledVar("myAuxVar")` and `_myAuxVar(coupled("myAuxVar"))` in the class definition and initialiser list in the .C file.

**Reply by GiudGiud** [reply]
This is the index of the variable.
You need to use coupledValue() to get the value at the local qp/node (for a nodal variable)

**Reply by Eilloo** [reply]
Thank you for clarifying.

I believe I do want the index of my aux variables, rather than the value in order to subsequently get the value at some other node, besides the local node/qp.

Printing the value of the index, I am seeing '-1' and '-2' for my two auxvariables.

Attempting to use these in, say, `someNode->dof_number(_aux_sys.number(), myVarIndex, 0)` results in a segfault (Which I think is expected, given the index seems wrong).
As a check, replacing 'myVarIndex' with '0', runs with no segfault (although I don't know which variable this corresponds to).

I think there are two questions from this:
~~1) Is it possible to use coupledValue() to get the variable at node besides the local node/qp? This would allow me to take a different approach entirely.~~ EDIT: I thought this part had been answered previously, and found the thread - I believe the coupled interface only allows you to access local values?

2) Should the index be negative for auxvariables? If not, have I done something wrong in declaring these as described above?

Thanks

**Reply by GiudGiud** [reply, marked as answer]
1) coupledValue should only give the local value
2) i suspect not but I dont see how we handle auxiliary variables in this code
https://mooseframework.inl.gov/docs/doxygen/moose/classCoupleable.html#ae04c38300772a77cd4632168742eea80

so I think this needs a patch. 

Other threads about this
https://github.com/idaholab/moose/discussions/21907
your own comment here
https://github.com/idaholab/moose/discussions/22895#discussioncomment-4342538

**Comment by Eilloo** [comment]
Ah yes - using `unsigned int` rather than `int`, I get the max possible integer minus 1 or 2, so it's definitely hitting that last line from https://mooseframework.inl.gov/docs/doxygen/moose/classCoupleable.html#ae04c38300772a77cd4632168742eea80 as you suggested.

As a workaround in the meantime, it seems I can subtract the number returned from `numeric_limits....etc....max()` to get back to `var->number()`.
The comment in the code you linked to suggests that generally, this could cause bad jacobians. Are we safe to do this in the context of aux kernels, since these do not have a residual contribution?

Thanks for your help!

**Reply by GiudGiud** [reply]
yeah dont worry about the bad jacobian message, that s for another context

**Reply by Eilloo** [reply]
Noted, thanks!

**Comment by YKumar1999** [comment]
Hey @Eilloo  I was wondering how did you used your method for accessing variables at nodes near to boundary ? I am also trying to do the same for variables whose values are stored in current solution (global vector) and using DOF map to find their information. Can you please guide me or provide an example where you looked at for your case.

**Reply by GiudGiud** [reply]
did you try to use the coupleable interface first? This is the preferred way
https://github.com/idaholab/moose/blob/next/framework/include/interfaces/Coupleable.h

**Reply by YKumar1999** [reply]
Is this method for getting values of nodal variables or elemental variables ?

**Reply by YKumar1999** [reply]
I want to get value of Temperature (Nodal variables) for local nodes in my mesh and also for non local nodes ?

**Reply by GiudGiud** [reply]
it does either, depending on the type of loop the object is in

is this for. am material? a kernel?

**Reply by YKumar1999** [reply]
This is for kernel (heat conduction equation) for temperature variable

**Reply by GiudGiud** [reply]
kernels use quadrature point based loop, not node. So coupledValue would give the value of the variable at each quadrature point

**Reply by YKumar1999** [reply]
Oh this was not aware of, do you have any idea, for the case of nodes. What method I can use to get the value of Temperature

**Reply by GiudGiud** [reply]
why do you need nodal values?

**Reply by YKumar1999** [reply]
So basically I am using XFEM for a moving interface in my system. In XFEM, moving interface is defined by a level set based user object. I need to get the nodal values like T present on each node ID, so that I can customize the level set user object based on the information regarding Temperature on nodes so that XFEM can access those nodal values and track the moving interface accordingly. For that reason first step I am struggling with is to get the global solution vector. I came across this forum I thought with this method I can access nodal values

**Reply by GiudGiud** [reply]
if you know the node you need the value on, you can do node->dof_number(system number, variable number, 0)
to get the index into the solution vector

**Reply by YKumar1999** [reply]
Guillaume thanks a lot. But the problem is I don't know the node ID or dof number or anything else. Basically what I know is that on my pellet inner boundary, axial nodes which are present I want the value on those nodes.

**Reply by YKumar1999** [reply]
This may be a dumb question: whatever we discussed in this forum like commands and all do I have to execute it via terminal or have to add those in some aux kernel .c or .h file or any other custom user object file ?

**Reply by GiudGiud** [reply]
> all do I have to execute it via terminal or have to add those in some aux kernel .c or .h file or any other custom user object file ?

depends if the object you need already exists or not

**Comment by YKumar1999** [comment]
dof_id_type ls_dof_id_1 = node1->dof_number(_system.number(), _level_set_var_number, 0);
        Number ls_node_1 = _solution(ls_dof_id_1);
        If you see here these syntax are already there in my level set user object of XFEM  for example. solution refers to global solution vector. but actually my main question was how can I get the global solution vector

**Reply by GiudGiud** [reply]
you get the solution vector usually from doing ` _solution(_sys.solution()),` in the constructor's initializer's list

you get _sys either because it already exists as an attribute or you can get it from any variable with variable.sys()

**Comment by YKumar1999** [comment]
Thanks Guillaume for your help, have understood now. Will try it out
